source/enum_set.h

```diff
@@ -12,195 +12,280 @@
+  }
 
-    if (a.overflow_ == nullptr && b.overflow_ == nullptr) {
-      return true;
+  // Removes the value `value` into the set.
+  // The set is unchanged if the value is not in the set.
+  void Remove(T value) {
+    const size_t index = FindBucketForValue(value);
+    if (index >= buckets_.size() ||
+        buckets_[index].start != ComputeBucketStart(value)) {
+      return;
+    }
+    auto& bucket = buckets_[index];
+    bucket.data &= ~compute_mask_for_value(value);
+    if (bucket.data == 0) {
+      buckets_.erase(buckets_.cbegin() + index);
     }
+  }
 
-    if (a.overflow_ == nullptr || b.overflow_ == nullptr) {
+  // Returns true if `value` is present in the set.
+  bool Contains(T value) const {
+    const size_t index = FindBucketForValue(value);
+    if (index >= buckets_.size() ||
+        buckets_[index].start != ComputeBucketStart(value)) {
       return false;
     }
-
-    return *a.overflow_ == *b.overflow_;
+    auto& bucket = buckets_[index];
+    return bucket.data & compute_mask_for_value(value);
   }
 
-  friend bool operator!=(const EnumSet& a, const EnumSet& b) {
-    return !(a == b);
+  // Calls `unaryFunction` once for each value in the set.
+  // Values are sorted in increasing order using their numerical values.
+  void ForEach(std::function<void(T)> unaryFunction) const {
+    for (const auto& bucket : buckets_) {
+      for (uint8_t i = 0; i < kBucketSize; i++) {
+        if (bucket.data & (1ULL << i)) {
+          unaryFunction(GetValueFromBucket(bucket, i));
+        }
+      }
+    }
   }
 
-  // Adds the given enum value to the set.  This has no effect if the
-  // enum value is already in the set.
-  void Add(EnumType c) { AddWord(ToWord(c)); }
+  // Returns true if the set is holds no values.
+  bool IsEmpty() const { return buckets_.size() == 0; }
 
-  // Removes the given enum value from the set.  This has no effect if the
-  // enum value is not in the set.
-  void Remove(EnumType c) { RemoveWord(ToWord(c)); }
+  // Returns true if this set contains at least one value contained in `in_set`.
+  // Note: If `in_set` is empty, this function returns true.
+  bool HasAnyOf(const EnumSet<T>& in_set) const {
+    if (in_set.IsEmpty()) {
+      return true;
+    }
 
-  // Returns true if this enum value is in the set.
-  bool Contains(EnumType c) const { return ContainsWord(ToWord(c)); }
+    for (auto& lhs_bucket : buckets_) {
+      for (auto& rhs_bucket : in_set.buckets_) {
+        if (lhs_bucket.start != rhs_bucket.start) {
+          continue;
+        }
 
-  // Applies f to each enum in the set, in order from smallest enum
-  // value to largest.
-  void ForEach(std::function<void(EnumType)> f) const {
-    for (uint32_t i = 0; i < 64; ++i) {
-      if (mask_ & AsMask(i)) f(static_cast<EnumType>(i));
-    }
-    if (overflow_) {
-      for (uint32_t c : *overflow_) f(static_cast<EnumType>(c));
+        if (lhs_bucket.data & rhs_bucket.data) {
+          return true;
+        }
+      }
```

Comment:
This is fine for now, but since the buckets are sorted, you could do an algorithm more similar to a merge algorithm.

```
rhs = buckets_.begin();
lhs = in_set.buckets_.end();
while (rhs != buckets_.end() && lhs != in_set.end()) {
  if (rhs->start == lhs->start) {
    if (rhs->data & lhs->data) return true;
    rhs++;
    lhs++;
  } else if (rhs->start < lhs->start) {
    rhs++;
  } else {
    lhs++;
  }
}
return false;
```

Reply:
That's a valid point, as yes, it'd be better. Now isn't that going against the other comment asking to use lower_bound only (as a more "complex" algorithm also means more possibilities for bugs)?

If you are fine with the added complexity, sure, I can change it, otherwise this version has the advantage to be dead-simple.

Reply:
As discussed, moving to this, complexity improvement is more important than simplicity here.